src/manager/checkpoint/subsystem.rs

```diff
@@ -0,0 +1,202 @@
+// Copyright 2022-2023 Protocol Labs
+// SPDX-License-Identifier: MIT
+use crate::config::{ReloadableConfig, Subnet};
+use crate::manager::checkpoint::manager::CheckpointManager;
+use crate::manager::checkpoint::submit::SubmissionStrategy;
+use anyhow::{anyhow, Result};
+use async_trait::async_trait;
+use futures_util::StreamExt;
+use fvm_shared::address::Address;
+use ipc_sdk::subnet_id::SubnetID;
+use std::collections::HashMap;
+use std::sync::Arc;
+use std::time::{Duration, Instant};
+use tokio::select;
+use tokio::sync::broadcast;
+use tokio::time::sleep;
+use tokio_graceful_shutdown::{IntoSubsystem, SubsystemHandle};
+
+const EXECUTION_BATCH_SIZE: usize = 20;
+const TASKS_PROCESS_THRESHOLD_SEC: u64 = 15;
+
+pub struct CheckpointSubsystem {
+    driver: CheckpointDriver,
+}
+
+#[async_trait]
+impl IntoSubsystem<anyhow::Error> for CheckpointSubsystem {
+    async fn run(self, subsys: SubsystemHandle) -> anyhow::Result<()> {
+        select! {
+            // TODO: somehow adding the below lines causes: higher-ranked lifetime error
+            // TODO: comment off for now first.
+            // _ = tokio::spawn(async move {
+            //     self.driver.run().await;
+            // }) => Ok::<(), anyhow::Error>(()),
+            _ = subsys.on_shutdown_requested() => {
+                log::info!("Shutting down checkpointing subsystem");
+                Ok(())
+            }
+        }
+    }
+}
+
+struct CheckpointDriver {
```

Comment:
I don't get why need so many abstractions, they are super confusing: 
CheckpointSubsystem --> CheckpointDriver --> CheckpointManagerWrapper --> Tokio tasks

Can't we simplify it a bit by just having a CheckpointSubsystems that spawns the CheckpointManagers for each of the checkpoints and that's it?

Reply:
already removed